cezo_fl/server.py

```diff
@@ -161,23 +166,29 @@ def train_one_step(self, iteration: int) -> tuple[float, float]:
         local_grad_scalar_list: list[list[torch.Tensor]] = []  # Clients X Local_update
         step_train_loss = Metric("Step train loss")
         step_train_accuracy = Metric("Step train accuracy")
-        for index in sampled_client_index:
-            client = self.clients[index]
-            last_update_iter = self.client_last_updates[index]
-            # The seed and grad in last_update_iter is fetched as well
-            # Note at that iteration, we just reset the client model so that iteration
-            # information is needed as well.
-            seeds_list = self.seed_grad_records.fetch_seed_records(last_update_iter)
-            grad_list = self.seed_grad_records.fetch_grad_records(last_update_iter)
-            # client will reset model to last pull states before update its model to match server
-            client.pull_model(seeds_list, grad_list)
-
-            client_local_update_result = client.local_update(seeds=seeds)
 
+        def client_work(index):
+            with torch.no_grad():
+                client = self.clients[index]
+                last_update_iter = self.client_last_updates[index]
+                # The seed and grad in last_update_iter is fetched as well
+                # Note at that iteration, we just reset the client model so that iteration
+                # information is needed as well.
+                seeds_list = self.seed_grad_records.fetch_seed_records(last_update_iter)
+                grad_list = self.seed_grad_records.fetch_grad_records(last_update_iter)
+                # client will reset model to last pull state before update its model to match server
+                client.pull_model(seeds_list, grad_list)
+
+                client_local_update_result = client.local_update(seeds=seeds).to(self.device)
+            return client_local_update_result
```

Comment:
better to make this function having the interface between client and server only. It will make the future multi-machine RPC refactor easy.

```python
def client_job( client: AbstractClient, seeds_list: ..., grad_list: ..) ->list[torch.tensor]:
    client.pull_model(seeds_list, grad_list)
    client_local_update_result = client.local_update(seeds=seeds)
    return client_local_update_result
```
Also, whenever it is possible, it is better to make this function pure functional instead of a closure of server's method.

I feel the `.to(self.device)` should be happened in the server side as well. But if this degrade the performance,  we can worry about it latter.

Reply:
refactored as suggested